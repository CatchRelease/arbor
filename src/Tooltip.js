import React from 'react';
import { injectGlobal } from 'react-emotion';
import { withTheme } from 'emotion-theming';
import PropTypes from 'prop-types';
import 'react-tippy/dist/tippy.css';
import { Tooltip as TippyTooltip } from 'react-tippy';

import Text from './Text';

// eslint-disable-next-line no-unused-expressions
injectGlobal`
  .tippy-tooltip.arbor-theme {
    padding: 0;
    border-radius: 3px;
  }
`;

export const Tooltip = ({ text, theme, children, ...props }) => {
  const tooltipText = (
    <Text
      px="smaller"
      py="smallest"
      theme={theme}
      fontSize="size2"
      color="white"
    >
      {text}
    </Text>
  );

  return (
    <TippyTooltip
      {...{
        ...props,
        theme: 'arbor',
        arrow: true,
        html: tooltipText,
        distance: 8,
        duration: 300,
        updateDuration: 0,
        arrowSize: 'small',
        animation: 'fade',
        inertia: true,
        animateFill: false
      }}
    >
      {children}
    </TippyTooltip>
  );
};

Tooltip.propTypes = {
  /**
   * Text to display within the tooltip when it is displayed
   * */
  text: PropTypes.string.isRequired,

  /**
   * Node which will trigger the tooltip. This should be either an Icon, Button,
   * or Link.
   * */
  children: PropTypes.element.isRequired
};

export default withTheme(Tooltip);
