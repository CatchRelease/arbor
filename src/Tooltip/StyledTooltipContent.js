import Box from '../Box';

const StyledTooltipContent = Box;

StyledTooltipContent.defaultProps = {
  bg: 'monochrome.grey90',
  borderRadius: 'small',
  color: 'white',
  fontSize: 'size4',
  p: 'smaller'
};

export default StyledTooltipContent;
