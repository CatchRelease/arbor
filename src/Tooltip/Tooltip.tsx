import { ComponentProps, ReactNode, useState, VFC } from 'react';
import { css } from '@emotion/react';
import Tippy from '@tippyjs/react/headless';
import { sticky } from 'tippy.js';
import { motion } from 'framer-motion';

import Box from '../Box';
import Card from '../Card';
import Grid from '../Grid';
import Heading from '../Heading';
import Text from '../Text';

type TippyProps = ComponentProps<typeof Tippy>;

type Props = {
  children: TippyProps['children'];
  content?: ReactNode;
  title?: ReactNode;
};

type TippyRenderAttributes = Parameters<
  Exclude<TippyProps['render'], undefined>
>[0];

const Tooltip: VFC<Props> = ({ children, content, title, ...props }) => {
  const [mounted, setMounted] = useState(false);

  const lazyPlugin = {
    fn: () => ({
      onMount: () => setMounted(true),
      onHide: () => setMounted(false)
    })
  };

  const springConfig = {
    type: 'spring',
    damping: 24,
    stiffness: 480,
    restSpeed: 0.1,
    restDelta: 0.6
  };

  const variants = {
    visible: { opacity: 1, scale: 1 },
    hidden: { opacity: 0, scale: 0.5 }
  };

  const renderTooltip = (attrs: TippyRenderAttributes) => (
    <Box
      as={motion.div}
      initial="hidden"
      variants={variants}
      animate={mounted ? 'visible' : 'hidden'}
      transition={springConfig}
      {...attrs}
    >
      <Card
        bg="monochrome.grey90"
        borderRadius="small"
        boxShadow="elevation1"
        color="monochrome.white"
        gridGap="smallest"
        fontSize="size3"
        maxWidth="200px"
        p="smaller"
        css={css`
          word-break: break-word;
        `}
        {...props}
      >
        {title && (
          <Heading.H3 color="inherit" mb="0">
            {title}
          </Heading.H3>
        )}
        {content && (
          <Text
            as={Grid}
            fontSize="inherit"
            color="inherit"
            textAlign="inherit"
          >
            {content}
          </Text>
        )}
      </Card>
    </Box>
  );

  return (
    <Tippy
      render={(attrs) => (mounted ? renderTooltip(attrs) : '')}
      offset={[0, 4]}
      plugins={[sticky, lazyPlugin]}
      popperOptions={{
        modifiers: [
          {
            name: 'preventOverflow',
            options: {
              padding: 4
            }
          }
        ]
      }}
    >
      {children}
    </Tippy>
  );
};

export default Tooltip;
