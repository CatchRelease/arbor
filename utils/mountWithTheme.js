import { mount } from 'enzyme';
import { ThemeProvider } from '@emotion/react';
import theme from '../src/theme';

export default (Component, mountOptions = {}) => {
  const wrappedComponent = (
    <ThemeProvider theme={theme}>{Component}</ThemeProvider>
  );

  const mounted = mount(wrappedComponent, mountOptions);

  return mounted.find(Component.type);
};
