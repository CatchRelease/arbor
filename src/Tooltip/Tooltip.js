/** @jsx jsx */
import React from 'react'; /* eslint-disable-line no-unused-vars */
import { Global, css, jsx } from '@emotion/core';
import { ThemeProvider, withTheme } from 'emotion-theming';
import PropTypes from 'prop-types';
import Tippy from '@tippyjs/react';
import 'tippy.js/dist/tippy.css';
import { animateFill, sticky } from 'tippy.js';

import StyledTooltipContent from './StyledTooltipContent';

const globalTippyStyles = (theme) => {
  const arrowColor = theme.colors.monochrome.grey90;

  return css`
    .tippy-popper {
      .tippy-tooltip.arbor-theme {
        background: none;
        padding: 0;
      }
    }

    .tippy-popper[x-placement='bottom'] .tippy-tooltip.arbor-theme [x-arrow] {
      border-bottom-color: ${arrowColor};
    }

    .tippy-popper[x-placement='top'] .tippy-tooltip.arbor-theme [x-arrow] {
      border-top-color: ${arrowColor};
    }

    .tippy-popper[x-placement='left'] .tippy-tooltip.arbor-theme [x-arrow] {
      border-left-color: ${arrowColor};
    }

    .tippy-popper[x-placement='right'] .tippy-tooltip.arbor-theme [x-arrow] {
      border-right-color: ${arrowColor};
    }
  `;
};

const Tooltip = ({ content, children, theme, position, ...props }) => {
  const tooltipContent = (
    <ThemeProvider theme={theme}>
      <StyledTooltipContent>{content}</StyledTooltipContent>
    </ThemeProvider>
  );

  return (
    <>
      <Global styles={globalTippyStyles} />

      <Tippy
        {...{
          theme: 'arbor',
          arrow: true,
          content: tooltipContent,
          duration: 300,
          animation: 'fade',
          inertia: true,
          plugins: [animateFill, sticky],
          placement: position,
          ...props
        }}
      >
        {children}
      </Tippy>
    </>
  );
};

Tooltip.propTypes = {
  /**
   * Node which will trigger the tooltip. This should be either an Icon, Button,
   * or Link.
   * */
  children: PropTypes.element.isRequired,

  /**
   * Content to display within the tooltip when it is displayed
   * */
  content: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,

  /**
   * Position for the tooltip (deprecated, use placement instead)
   * */
  position: PropTypes.string,

  /**
   * Theme used for styling the Tooltip.
   */
  theme: PropTypes.object.isRequired // eslint-disable-line react/forbid-prop-types
};

Tooltip.defaultProps = {
  position: 'top'
};

export default withTheme(Tooltip);
