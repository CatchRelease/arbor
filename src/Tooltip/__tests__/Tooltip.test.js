import React from 'react';

import mountWithTheme from '../../../utils/mountWithTheme';
import Tooltip from '../Tooltip';

describe('<Tooltip />', () => {
  it('properly renders a tooltip', () => {
    const tree = mountWithTheme(
      <Tooltip content="Hello World">
        <p>I have a tooltip</p>
      </Tooltip>
    );

    expect(tree).toMatchSnapshot();
  });
});
