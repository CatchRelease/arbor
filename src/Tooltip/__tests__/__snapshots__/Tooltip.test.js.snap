// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> properly renders a tooltip 1`] = `
<WithTheme(Tooltip)
  text="Hello World"
>
  <Tooltip
    text="Hello World"
    theme={
      Object {
        "backgroundColors": Object {
          "default": "#F2F2F2",
          "muted": "#F8F9F9",
        },
        "borderColors": Object {
          "default": "#E0E1E2",
          "muted": "#EBECEC",
        },
        "borderRadius": Object {
          "large": "6px",
          "small": "3px",
        },
        "borderWidth": Object {
          "large": "4px",
          "medium": "2px",
          "small": "1px",
        },
        "brandColors": Object {
          "dark": "#00533E",
          "darkest": "#001C15",
          "default": "#007558",
          "light": "rgba(0,83,62,0.09)",
        },
        "brandFont": "Motiva Sans",
        "breakpoints": Array [
          "512px",
          "768px",
          "1024px",
          "1600px",
        ],
        "colors": Object {
          "background": Object {
            "default": "#F2F2F2",
            "muted": "#F8F9F9",
          },
          "black": "#000000",
          "blue": "#4466EE",
          "blueDark": "#4455CC",
          "blueDarkest": "#2E374E",
          "blueLight": "#5577FF",
          "blueLighter": "#6688FF",
          "border": Object {
            "default": "#E0E1E2",
            "muted": "#EBECEC",
          },
          "brand": Object {
            "dark": "#00533E",
            "darkest": "#001C15",
            "default": "#007558",
            "light": "rgba(0,83,62,0.09)",
          },
          "bronze": "#CC7700",
          "bronzeDark": "#AA5500",
          "bronzeLight": "#CC9900",
          "green": "#00AA44",
          "greenDark": "#008833",
          "greenLight": "#00BB44",
          "greenLighter": "#00CC66",
          "grey10": "#F8F9F9",
          "grey100": "#0F0F10",
          "grey20": "#F2F2F2",
          "grey30": "#EBECEC",
          "grey40": "#E0E1E2",
          "grey50": "#BEBEC0",
          "grey60": "#989A9C",
          "grey70": "#78797D",
          "grey80": "#5B5D61",
          "grey90": "#25272D",
          "icon": Object {
            "default": "#989A9C",
            "disabled": "rgba(190,190,192,0.7)",
          },
          "red": "#EE2200",
          "redDark": "#DD1100",
          "redLight": "#FF4444",
          "redLighter": "#FF6655",
          "text": Object {
            "dark": "#0F0F10",
            "default": "#25272D",
            "disabled": "rgba(120,121,125,0.7)",
            "muted": "#5B5D61",
          },
          "white": "#FFFFFF",
        },
        "fontSizes": Object {
          "size1": "0.6875rem",
          "size2": "0.75rem",
          "size3": "0.8125rem",
          "size4": "0.875rem",
          "size5": "1rem",
          "size6": "1.25rem",
          "size7": "1.5rem",
          "size8": "1.75rem",
          "size9": "2.125rem",
        },
        "fontWeights": Object {
          "bold": 700,
          "medium": 500,
          "regular": 400,
        },
        "iconColors": Object {
          "default": "#989A9C",
          "disabled": "rgba(190,190,192,0.7)",
        },
        "iconFontPrefix": "ar",
        "lineHeights": Object {
          "large": 1.5,
          "small": 1.25,
        },
        "shadows": Object {
          "elevation0": "0 0 1px rgba(37,39,45,0.47)",
          "elevation1": "0 0 1px rgba(37,39,45,0.47), 0 2px 4px -2px rgba(37,39,45,0.3)",
          "elevation2": "0 0 1px rgba(37,39,45,0.47), 0 5px 8px -4px rgba(37,39,45,0.3)",
          "elevation3": "0 0 1px rgba(37,39,45,0.47), 0 8px 10px -4px rgba(37,39,45,0.3)",
          "elevation4": "0 0 1px rgba(37,39,45,0.47), 0 16px 24px -8px rgba(37,39,45,0.3)",
        },
        "space": Object {
          "large": "1.5rem",
          "larger": "2rem",
          "largest": "4rem",
          "regular": "1rem",
          "small": "0.75rem",
          "smaller": "0.5rem",
          "smallest": "0.25rem",
        },
        "textColors": Object {
          "dark": "#0F0F10",
          "default": "#25272D",
          "disabled": "rgba(120,121,125,0.7)",
          "muted": "#5B5D61",
        },
        "typography": Object {
          "text": Object {
            "longForm": Object {
              "fontSize": "1rem",
            },
            "tiny": Object {
              "fontSize": "0.75rem",
            },
            "ui": Object {
              "fontSize": "0.875rem",
            },
          },
        },
      }
    }
  >
    <ForwardRef(render)
      styles={
        Object {
          "map": undefined,
          "name": "957k5c-globalTippyStyles",
          "next": undefined,
          "styles": ".tippy-tooltip.arbor-theme{border-radius:3px;padding:0;}label:globalTippyStyles;",
        }
      }
    >
      <InnerGlobal
        cache={
          Object {
            "insert": [Function],
            "inserted": Object {},
            "key": "css",
            "nonce": undefined,
            "registered": Object {},
            "sheet": StyleSheet {
              "before": null,
              "container": <head>
                <style
                  data-emotion="css-global"
                >
                  
                  .tippy-tooltip.arbor-theme{border-radius:3px;padding:0;}
                </style>
              </head>,
              "ctr": 0,
              "isSpeedy": false,
              "key": "css",
              "nonce": undefined,
              "tags": Array [],
            },
          }
        }
        serialized={
          Object {
            "map": undefined,
            "name": "957k5c-globalTippyStyles",
            "next": undefined,
            "styles": ".tippy-tooltip.arbor-theme{border-radius:3px;padding:0;}label:globalTippyStyles;",
          }
        }
      />
    </ForwardRef(render)>
    <Tooltip
      animateFill={false}
      animation="fade"
      arrow={true}
      arrowSize="small"
      className=""
      delay={0}
      disabled={false}
      distance={8}
      duration={300}
      followCursor={false}
      hideDelay={0}
      hideDuration={375}
      hideOnClick={true}
      html={
        <ForwardRef(render)
          color="white"
          fontSize="size4"
          px="smaller"
          py="smallest"
          theme={
            Object {
              "backgroundColors": Object {
                "default": "#F2F2F2",
                "muted": "#F8F9F9",
              },
              "borderColors": Object {
                "default": "#E0E1E2",
                "muted": "#EBECEC",
              },
              "borderRadius": Object {
                "large": "6px",
                "small": "3px",
              },
              "borderWidth": Object {
                "large": "4px",
                "medium": "2px",
                "small": "1px",
              },
              "brandColors": Object {
                "dark": "#00533E",
                "darkest": "#001C15",
                "default": "#007558",
                "light": "rgba(0,83,62,0.09)",
              },
              "brandFont": "Motiva Sans",
              "breakpoints": Array [
                "512px",
                "768px",
                "1024px",
                "1600px",
              ],
              "colors": Object {
                "background": Object {
                  "default": "#F2F2F2",
                  "muted": "#F8F9F9",
                },
                "black": "#000000",
                "blue": "#4466EE",
                "blueDark": "#4455CC",
                "blueDarkest": "#2E374E",
                "blueLight": "#5577FF",
                "blueLighter": "#6688FF",
                "border": Object {
                  "default": "#E0E1E2",
                  "muted": "#EBECEC",
                },
                "brand": Object {
                  "dark": "#00533E",
                  "darkest": "#001C15",
                  "default": "#007558",
                  "light": "rgba(0,83,62,0.09)",
                },
                "bronze": "#CC7700",
                "bronzeDark": "#AA5500",
                "bronzeLight": "#CC9900",
                "green": "#00AA44",
                "greenDark": "#008833",
                "greenLight": "#00BB44",
                "greenLighter": "#00CC66",
                "grey10": "#F8F9F9",
                "grey100": "#0F0F10",
                "grey20": "#F2F2F2",
                "grey30": "#EBECEC",
                "grey40": "#E0E1E2",
                "grey50": "#BEBEC0",
                "grey60": "#989A9C",
                "grey70": "#78797D",
                "grey80": "#5B5D61",
                "grey90": "#25272D",
                "icon": Object {
                  "default": "#989A9C",
                  "disabled": "rgba(190,190,192,0.7)",
                },
                "red": "#EE2200",
                "redDark": "#DD1100",
                "redLight": "#FF4444",
                "redLighter": "#FF6655",
                "text": Object {
                  "dark": "#0F0F10",
                  "default": "#25272D",
                  "disabled": "rgba(120,121,125,0.7)",
                  "muted": "#5B5D61",
                },
                "white": "#FFFFFF",
              },
              "fontSizes": Object {
                "size1": "0.6875rem",
                "size2": "0.75rem",
                "size3": "0.8125rem",
                "size4": "0.875rem",
                "size5": "1rem",
                "size6": "1.25rem",
                "size7": "1.5rem",
                "size8": "1.75rem",
                "size9": "2.125rem",
              },
              "fontWeights": Object {
                "bold": 700,
                "medium": 500,
                "regular": 400,
              },
              "iconColors": Object {
                "default": "#989A9C",
                "disabled": "rgba(190,190,192,0.7)",
              },
              "iconFontPrefix": "ar",
              "lineHeights": Object {
                "large": 1.5,
                "small": 1.25,
              },
              "shadows": Object {
                "elevation0": "0 0 1px rgba(37,39,45,0.47)",
                "elevation1": "0 0 1px rgba(37,39,45,0.47), 0 2px 4px -2px rgba(37,39,45,0.3)",
                "elevation2": "0 0 1px rgba(37,39,45,0.47), 0 5px 8px -4px rgba(37,39,45,0.3)",
                "elevation3": "0 0 1px rgba(37,39,45,0.47), 0 8px 10px -4px rgba(37,39,45,0.3)",
                "elevation4": "0 0 1px rgba(37,39,45,0.47), 0 16px 24px -8px rgba(37,39,45,0.3)",
              },
              "space": Object {
                "large": "1.5rem",
                "larger": "2rem",
                "largest": "4rem",
                "regular": "1rem",
                "small": "0.75rem",
                "smaller": "0.5rem",
                "smallest": "0.25rem",
              },
              "textColors": Object {
                "dark": "#0F0F10",
                "default": "#25272D",
                "disabled": "rgba(120,121,125,0.7)",
                "muted": "#5B5D61",
              },
              "typography": Object {
                "text": Object {
                  "longForm": Object {
                    "fontSize": "1rem",
                  },
                  "tiny": Object {
                    "fontSize": "0.75rem",
                  },
                  "ui": Object {
                    "fontSize": "0.875rem",
                  },
                },
              },
            }
          }
        >
          Hello World
        </ForwardRef(render)>
      }
      inertia={true}
      interactive={false}
      interactiveBorder={2}
      multiple={false}
      offset={0}
      onHidden={[Function]}
      onHide={[Function]}
      onRequestClose={[Function]}
      onShow={[Function]}
      onShown={[Function]}
      popperOptions={Object {}}
      position="top"
      size="regular"
      sticky={false}
      stickyDuration={200}
      style={Object {}}
      theme="arbor"
      touchHold={false}
      trigger="mouseenter focus"
      unmountHTMLWhenHide={false}
      updateDuration={0}
    >
      <div
        className=""
        style={
          Object {
            "display": "inline",
          }
        }
      >
        <p>
          I have a tooltip
        </p>
      </div>
    </Tooltip>
  </Tooltip>
</WithTheme(Tooltip)>
`;
